pep-0639.rst

```diff
@@ -1,543 +1,2208 @@
+more obviously the preferred approach. However, it is far *less* obvious that
+it is a license expression at all, to authors and those viewing the files,
+and this lack of clarity, explicitness, ambiguity and potential for user
+confusion is exactly what this PEP seeks to avoid, all to save a few characters
+over other approaches.
+
+If an ``expression`` subkey was added to the ``license`` table, it would retain
+the clarity of a new top-level key, but add additional complexity for no
+real benefit, with an extra level of nesting, and users and tools needing to
+deal with the mutual exclusivity of the subkeys, as before. And allowing both
+(as a table subkey *and* the string value) would inherit both's downsides,
+while adding even more spec and tool complexity and making there more than
+"one obvious way to do it", further potentially confusing users.
+
+Therefore, a separate top-level ``license-expression`` key was adopted to avoid
+all these issues, with relatively minimal downside aside from adding a single
+additional top-level key and (versus some approaches) a few extra characters
+to type.
+
+
+Add a ``type`` key to treat as expression
+'''''''''''''''''''''''''''''''''''''''''
+
+Instead of creating a new top-level ``license-expression`` key in the
+PEP 621 source metadata, one could add a ``type`` subkey to the existing
+``license`` table to control whether ``text`` (or a string value)
+is interpreted as free-text or a license expression. This could make
+backward compatibility a little more seamless, as older tools could ignore
+it and always treat ``text`` as ``license``, while newer tools would
+know to treat it as a license expression, if ``type`` was set appropriately.
+Indeed, PEP 621 seems to suggest something of this sort as a possible
+alternative way that SPDX license expressions could be implemented.
+
+However, all the same downsides as in the previous item apply here,
+including greater complexity, a more complex mapping between the project
+source metadata and core metadata and inconsistency between the presentation
+in tool config, PEP 621 and core metadata, a much less clean deprecation,
+further bikeshedding over what to name it, and inability to mark one but
+not the other as dynamic, among others.
+
+In addition, while theoretically potentially a little easier in the short
+term, in the long term it would mean users would always have to remember
+to specify the correct ``type`` to ensure their license expression is
+interpreted correctly, which adds work and potential for error; we could
+never safety change the default while being confident that users
+understand that what they are entering is unambiguously a license expression,
+with all the false positive and false negative issues as above.
+
+Therefore, for these as well as the same reasons this approach was rejected
+for the core metadata in favor of a distinct ``License-Expression`` field,
+we similarly reject this here.
+
+
+Must be marked dynamic to back-fill
```

Comment:
Doesn't PEP 621 mandate that the field be marked as dynamic to be backfilled?

> Only when a field is marked as dynamic may a tool provide a "new" value.

* https://www.python.org/dev/peps/pep-0621/#specification
* https://www.python.org/dev/peps/pep-0621/#have-the-dynamic-field-only-require-specifying-missing-required-fields

Or am I misunderstanding what you're saying here?

Reply:
Potentially, depending on how strictly it is interpreted, but I address this explicitly in some detail in this very section. Given it is simply using the verbatim value of another specific, mutually-exclusive key as defined by this specification, such that any tool reading the `pyproject.toml` can "trivially, deterministically and unambiguously determine this using only the static data in the ``pyproject.toml`` file itself," it meets the spirit and intent, if (arguably) not the exact letter of PEP 621.

While I'd normally argue for explicitness, marking the field as dynamic has serious practical danger, as it means that any tool (including those not yet unaware of this PEP) could fill it with anything it wanted, directly violating the specification in this PEP, which is why the specification here explicitly _prohibits_ marking `license` as `dynamic` if `license-expression` is also defined.

Also, as a side note, requiring it to be dynamic is unavoidably mutually exclusive with @uranusjr and @ofek 's proposal to move `license-expression` back under the `license` key, since it would be impossible to mark one one as dynamic without the other, and thus prohibited to define a license expression value there in the first place.

In any case, as this PEP updates the specification originally defined in PEP 621, it does not irrevocably constrain this PEP (in fact, for `license-files`, I also had to make what was arguably another mild exception in order to maintain safe, sane and backward-compatible default behavior, as also discussed here), and this section of the canonical project metadata spec will be updated accordingly to clarify these edge cases.

That said, I'm not completely wed to the necessity of allowing back-filling in the first place, as I haven't yet seen a 100% compelling rationale either way other than general backward compat and people on Discourse asking for it.